### ISSUE TITLE
[2021] GPT Understands, Too

### ISSUE THREAD

**Issue by cfiken** [cfiken/paper-reading#124]
## 0. 論文

| タイトル | GPT Understands, Too |
|:-------|:--|
| リンク  | https://arxiv.org/abs/2103.10385 , https://github.com/THUDM/P-tuning |
| 著者    | Xiao Liu, Yanan Zheng, Zhengxiao Du, Ming Ding, Yujie Qian, Zhilin Yang, Jie Tang |
| 投稿日付 | 2021/3/18 on arXiv |

## 1. どんなもの？

GPT 系モデルは、生成系タスクにおける目覚ましい活躍の一方、GLUE のような NLU 系タスクではスコアがいまひとつだった。 本研究では NLU 系の下位タスクへの適合に従来の finetuning による方法ではなく、モデルへの入力である prompt を最適化する P-tuning を提案。GPT 系モデルでの NLU タスクでのスコアを大幅に改善した。
同様の方法は BERT 系モデルにも効果があり、few-shot の SuperGLUE で ALBERT を用いて SoTA を更新。

![スクリーンショット 2021-03-23 0 32 11](https://user-images.githubusercontent.com/1255485/112015512-39a9eb80-8b6f-11eb-8306-63ab1972a81e.png)

## 2. 先行研究と比べてどこがすごい？

- finetuning に代わる downstream task への適合方法として P-tuning を提案。
- prompt の最適化は既存研究では discrete な扱いが主だったが、prompts 自体を連続値表現として扱うことでそのまま勾配計算し最適化を行えるようにした。

## 3. 技術や手法のキモはどこ？

![スクリーンショット 2021-03-23 0 18 31](https://user-images.githubusercontent.com/1255485/112013349-52190680-8b6d-11eb-8d03-cf80298ce2c0.png)

- 入力の prompt を p, context を x, target を y として、ひとまとめにしたものを template T とする。
    - 論文中の例としては、 `The capital of Britain is [MASK]` というテンプレートにおいて、 `Britain` が context, `[MASK]` が target, 他が prompt に当たる。
    - ![スクリーンショット 2021-03-23 0 38 47](https://user-images.githubusercontent.com/1255485/112016541-26e3e680-8b70-11eb-8891-913fa0bcf555.png)
- 従来手法では、template を学習済み言語モデルの embedding 層を通し、モデル本体への入力とする
    - 上図左の例。 `e(・)` が各トークンの embeddings を表す。embedding 後の表現は次のようになる。
    - ![スクリーンショット 2021-03-23 0 39 13](https://user-images.githubusercontent.com/1255485/112016633-3b27e380-8b70-11eb-9c4b-f0c57dbc5ee2.png)
- P-tuning では、prompts の部分は単語として embedding に通すのではなく、それ専用の学習可能なテンソル `h_i` を用いる
    - ![スクリーンショット 2021-03-23 0 40 33](https://user-images.githubusercontent.com/1255485/112016820-66aace00-8b70-11eb-8af7-b7e604738250.png)
    - これにより、辞書内の単語を用いるよりも柔軟に prompts を表現可能になる。
    - あとはタスク用のロスを定義して、`h_i` を最適化する。
        - ![スクリーンショット 2021-03-23 0 42 20](https://user-images.githubusercontent.com/1255485/112017101-a5408880-8b70-11eb-8d88-5749eb1d5e78.png)
    - prompts として最初からトークンを扱っていないため、直接最適化可能。
- `h_i` の生成はランダム初期化ではなく、元の embedding に対して1層の Bi-LSTM に2層の MLP で作った小さなモデルの出力を使用する
    - ![スクリーンショット 2021-03-23 0 44 01](https://user-images.githubusercontent.com/1255485/112017746-3f083580-8b71-11eb-974a-f08b089806bd.png)
    - [memo] 記載が無い気がしたけど初期入力はランダム?
    - こうすることにより、 (1) Discreteness (2) Association の問題が解決するとのこと
        - (1) pretrained model の embedding は既にそれぞれ discrete なので、ランダム初期化した h を使ってしまうとすぐ局所解に入ってしまう
        - (2) 理想的にはそれぞれの`h_i` は独立ではなく互いに相関があってほしい
        - [memo] (1), (2) については正直良く分からん
- 一部で特殊なトークンはそのまま単語として入れている
    - `?` だったり、都市を聞く際の `capital` だったり。

## 4. どうやって有効だと検証した？

Knowledge probing のベンチマークセットである LAMA と、SuperGLUE で実験。Finetuning (FT), Manual Prompt (MP), MP + FT, P-tuning の比較をメインにしている。Manual Prompt はデータセットで与えられる Prompt のこと。
LAMA では、P-tuning によって BERT 系モデル、 GPT 系モデルともに性能を改善できた。特に FT の恩恵が小さかった GPT 系モデルは P-Tuning で大きく性能を改善し、メモリの問題で finetune が難しい MegatoronLM を用いて P-Tuning で SoTA を更新した。

![スクリーンショット 2021-03-23 12 03 54](https://user-images.githubusercontent.com/1255485/112086610-e2883300-8bcf-11eb-8571-6ca0a4884aca.png)

SuperGLUE とその few-shot 版 FewGLUE の実験でも同様に、多くのタスクで P-Tuning が FT を上回る性能となるだけでなく、P-Tuning を用いた GPT モデルが同サイズの BERT のスコアを上回った。データ数が多い場合はFT に軍配が上がるケースが多く、P-Tuning は特にデータが少ない場合に有効であるとの知見が得られた。
また、few-shot の設定で ALBERT + P-Tuning でのスコアが SoTA を更新している。

![スクリーンショット 2021-03-23 12 05 38](https://user-images.githubusercontent.com/1255485/112086762-2bd88280-8bd0-11eb-8d29-d7d7facfe435.png)

他の prompt 生成系の手法と比較しても P-Tuning がスコアを改善している。

## 5. 議論はある？

今まで NLU タスクが苦手だとされていた GPT 系モデルにおいて、言語理解もできているという可能性が示された。
今後更に知識を抽出する方法や、内部表現の分析などの研究が進められそう。

## 6. 次に読むべき論文は？

- Language Models as Knowledge Bases?
    - Fabio Petroni, Tim Rocktäschel, Patrick Lewis, Anton Bakhtin, Yuxiang Wu, Alexander H. Miller, Sebastian Riedel (Facebook AI)
    - https://arxiv.org/abs/1909.01066
    - ベンチマークになっている LAMA の提案論文
- It's Not Just Size That Matters: Small Language Models Are Also Few-Shot Learners
    - Timo Schick, Hinrich Schütze
    - https://arxiv.org/abs/2009.07118
    - 比較モデルに使用されている、性能の良くなる prompt を探す + 性能の上がる手法まとめ的な研究
